### ISSUE TITLE
Undefined codes

### ISSUE THREAD

**Issue by aoern** [CatalogueOfLife/data#583]
There are 101.000+ low taxon names (species or infraspecific) and 64.000+ high taxon names without a defined nomenclatural Code.

Examples of datasets that totally lack nomenclatural codes:
Gymnodinium
Tortricid.net
WCVP-Fabaceae
Microsporidia
Strange enough, WCVP-Fabaceae however lists the codes of all subspecies taxa (1800+).

There are also several datasets that list the nomenclatural codes of low taxa but not of generic taxa and higher.

In most cases the correct code is easy to deduce, but because of missing codes it is impossible to design a general purpose script or program that needs to know the code, for example a species name syntax checker.

**Comment by mdoering** [comment]
There are 2 places to fix this. If an entire source dataset follows a single code we should declare that in the dataset settings and it would be applied to all its name upon the next _import_ - which we should then trigger.

Otherwise we can also set the code in every sector setting - which allows to apply a code also to mixed sources such as ITIS.

**Comment by yroskov** [comment]
It's not nice to learn that some GSDs have problems with Codes again. A year ago (or so) **all** GSDs in CoL have been associated with Codes. We'll have a look on the problem again.

Well, neither @gdower nor I have a control over **Tortricid.net** and **WCVP-Fabaceae**. These GSDs were converted and imported in CLB by other people. (@dhobern, could you please apply the ICZN to Tortricid.net during re-import?)

Whereas, **Gymnodinium** & **Microsporidia** have been moved in CLB from AC19. We'll see with Geoff, what we can do here.

**Comment by mdoering** [comment]
like I said above there are various ways to fix this, you don't need to change the archive files if you don't have access

**Comment by yroskov** [comment]
WCVP-Fabaceae already has "botanical' Code in the settings. 

![image](https://github.com/CatalogueOfLife/data/assets/28574252/34a2cc4c-c03c-4e1e-9312-a32e91943099)

I do re-sync now (2023-11-13)

**Comment by mdoering** [comment]
the dataset setting is for imports, not syncs...

**Comment by mdoering** [comment]
But the Fabaceae dataset does have the code imported already: https://www.checklistbank.org/dataset/2304/names?NOM_CODE=botanical

**Comment by yroskov** [comment]
@aoern, you gave 4 GSDs as an example. Do you have a full list of GSDs where Code is not defined?

**Comment by mdoering** [comment]
I only added code=zoological to the Tortricid.net dataset options and reimported it

**Comment by gdower** [comment]
@yroskov and I added code to the sectors for Gymnodinium and Microsporidia and resynced them.

**Comment by dhobern** [comment]
Thanks - should this property move into (or be duplicated in) the main metadata page/document? Or maybe the Options tab should be merged into the Metadata tab but only visible to those editing the page or to administrators. It could then be made a mandatory field for completion.

**Comment by mdoering** [comment]
That is how we started, but I feel it is cleaner to separate configurations/settings from metadata that informs you but is not used for interpretations. 

Apart from the nomenclatural code you can also declare extinct or the default environment in the settings. And to be honest I would not mind to be even more generic in the future and allow defaults for any coldp or dwc term just as we do with [default.yaml](https://github.com/CatalogueOfLife/coldp/blob/master/README.md#default-values). Maybe splitting default values from other settings helps to visualise them on the main metadata page?

**Comment by yroskov** [comment]
>allow defaults for any coldp or dwc term

I am supporting this idea. But: please use controlled vocabularies, where interface user have an option to chose value from the list.

**Comment by aoern** [comment]
@yroskov, apart from the 4 GSDs mentioned earlier (Gymnodinium, Tortricid.net, CVP-Fabaceae, and Microsporidia) there are no other datasets with specific and infraspecific taxa without a defined code.
However, there area still 56.000+ high rank taxa without a defined code in several datasets.

**Comment by DaveNicolson** [comment]
ICZN does regulate names above superfamily, just in fewer ways. So they still have to be properly published, but they are not subject to Priority, for example. Art. 1.2.2 notes the Articles that apply for names above the family-group ranks.

**Comment by aoern** [comment]
@yroskov, you write that "It looks like majority of them belong to the "management classification"". However, source 0 and IRMNG count only to 4.000+ cases and they are not included in these 56.000 cases.

**Comment by yroskov** [comment]
> source 0 and IRMNG count only to 4.000+ cases and they are not included in these 56.000 cases.

That is interesting. If these taxa are inside sectors, @mdoering, we need a list of parent GSDs where Code is not designated yet. There's not a quick way to get the info out of the API, as @gdower found.

**Comment by mdoering** [comment]
> Does this report http://api.checklistbank.org/dataset/3/nameusage/search?nomCode=_NULL&facet=rank&facet=SECTOR_DATASET_KEY have an answer to the question, which GSDs in CoL still have no Code assignment?

Yes, the facet tells you which source datasets (sectorDatasetKey) and ranks are involved. If you limit the search to zero you only get the facet to view: http://api.checklistbank.org/dataset/3/nameusage/search?nomCode=_NULL&facet=rank&facet=SECTOR_DATASET_KEY&limit=0

You can see that most come from [PBDB](https://www.checklistbank.org/dataset/268676/names?nomCode=_NULL).
Then [SF+](https://www.checklistbank.org/dataset/2073/names?nomCode=_NULL).

Btw, you can also add the filter `nomCode=_NULL` to the UI search URL and it uses it to filter even though the filter is not present in the forms yet (ping @thomasstjerne, see https://github.com/CatalogueOfLife/checklistbank/issues/1319).

> How to assign Code to the nodes in the management classification (there are no sectors there)?

You can define the code in the name editor. There is no bulk tool yet, but I am sure we'll need it.

**Comment by mdoering** [comment]
> How to assign Code to the nodes in the management classification (there are no sectors there)?

You can define the code in the name editor. There is no bulk tool yet, but I am sure we'll need it.
Well, checking the taxon forms there does not seem to be a code field:

<img width="888" alt="image" src="https://github.com/CatalogueOfLife/data/assets/327505/49992d0a-6f95-4212-bd36-880112a55cf4">

@thomasstjerne that would be need in the future. Also _extinct_ and really all the other name & taxon fields too, at least in an advanced section: https://github.com/CatalogueOfLife/checklistbank/issues/1320

**Comment by TonyRees** [comment]
If anything needs adjusting in IRMNG, I am happy to assist. I think
nomenclature Code is added via a script at the data export stage, it is not
in my edit interface.

**Comment by aoern** [comment]
More info about missing code GSDs:
There are tens of GSDs that contain taxa without nomenclatural code:

Species Fungorum Plus  14210 taxa
Systema Dipterorum  11456
FishBase  5758
WSC  4390
CilCat  1262
and many many more

**Comment by yroskov** [comment]
Well, Species Fungorum Plus, Systema Dipterorum, FishBase, WSC, CilCat, all have (and had before their imports) defined Code of Nomenclature in the Options. It means, we are doing Sisyphean labor. The problem is not in data, but in the CLB code.

**Comment by mdoering** [comment]
Ah, looking at [Fishbase examples](https://www.checklistbank.org/dataset/3/names?SECTOR_DATASET_KEY=1010&facet=rank&limit=50&nomCode=_NULL&offset=100&reverse=false) I see that these are all names not present directly as a record on it own in the source, but are names with [origin=denormalised](https://www.checklistbank.org/dataset/1010/taxon/xDC), i.e. they are found only in the flat, higher classification and need to be extracted. The code doing that probably does not add any default values

**Comment by gdower** [comment]
> The code doing that probably does not add any default values

Related: I was testing the default values and for ACEF imports with extinct set true in the dataset setting, the importer doesn't set the extinct flag on denormalized records.

**Comment by mdoering** [comment]
fixed in code now, but not will be deployed to prod only tomorrow: https://github.com/CatalogueOfLife/backend/commit/435d236048251c8c89a939a5cfb39146f003d22c
